File: src/data/transcript.ts
import { existsSync, readFileSync } from 'fs';

export interface TextTurn {
  kind: 'text';
  role: 'user' | 'assistant' | 'system';
  text: string;
}

export interface ToolCallTurn {
  kind: 'tool_call';
  id: string;
  name: string;
  input: string;
}

export interface ToolResultTurn {
  kind: 'tool_result';
  toolUseId: string;
  name: string;
  content: string;
}

export type Turn = TextTurn | ToolCallTurn | ToolResultTurn;

type ContentBlock =
  | { type: 'text'; text: string }
  | { type: 'tool_use'; id: string; name: string; input: unknown }
  | { type: 'tool_result'; tool_use_id: string; content: string | ContentBlock[] };

interface RawMessage {
  role?: string;
  content?: string | ContentBlock[];
}

function blockText(content: string | ContentBlock[] | undefined): string {
  if (!content) return '';
  if (typeof content === 'string') return content;
  return content
    .filter((b): b is { type: 'text'; text: string } => b.type === 'text')
    .map(b => b.text)
    .join('\n');
}

export function parseTranscript(filePath: string): Turn[] {
  if (!existsSync(filePath)) return [];

  try {
    const raw = readFileSync(filePath, 'utf-8');
    const jsonLines = raw.trim().split('\n').filter(l => l.trim());
    const turns: Turn[] = [];
    const toolNames = new Map<string, string>();

    for (const line of jsonLines) {
      try {
        const msg: RawMessage = JSON.parse(line);
        const role = msg.role;
        if (!role) continue;

        const content = msg.content;

        if (typeof content === 'string') {
          if (content.trim()) {
            turns.push({ kind: 'text', role: role as TextTurn['role'], text: content });
          }
          continue;
        }

        if (!Array.isArray(content)) continue;

        let pendingText = '';

        for (const block of content) {
          if (block.type === 'text') {
            if (block.text?.trim()) {
              pendingText += (pendingText ? '\n' : '') + block.text;
            }
          } else if (block.type === 'tool_use') {
            if (pendingText) {
              turns.push({ kind: 'text', role: role as TextTurn['role'], text: pendingText });
              pendingText = '';
            }
            toolNames.set(block.id, block.name);
            turns.push({
              kind: 'tool_call',
              id: block.id,
              name: block.name,
              input: JSON.stringify(block.input, null, 2),
            });
          } else if (block.type === 'tool_result') {
            if (pendingText) {
              turns.push({ kind: 'text', role: role as TextTurn['role'], text: pendingText });
              pendingText = '';
            }
            const resultContent = blockText(block.content);
            turns.push({
              kind: 'tool_result',
              toolUseId: block.tool_use_id,
              name: toolNames.get(block.tool_use_id) ?? 'unknown',
              content: resultContent,
            });
          }
        }

        if (pendingText) {
          turns.push({ kind: 'text', role: role as TextTurn['role'], text: pendingText });
        }
      } catch {
        // skip malformed lines
      }
    }

    return turns;
  } catch {
    return [];
  }
}

File: internal/data/transcript.go
package data

import (
	"bufio"
	"encoding/json"
	"os"
	"strings"
)

func ParseTranscript(filePath string) []Turn {
	f, err := os.Open(filePath)
	if err != nil {
		return nil
	}
	defer f.Close()

	var turns []Turn
	toolNames := map[string]string{}

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 4*1024*1024), 4*1024*1024)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		var msg struct {
			Role    string          `json:"role"`
			Content json.RawMessage `json:"content"`
		}
		if err := json.Unmarshal([]byte(line), &msg); err != nil || msg.Role == "" {
			continue
		}

		// Try string content
		var strContent string
		if err := json.Unmarshal(msg.Content, &strContent); err == nil {
			if strings.TrimSpace(strContent) != "" {
				turns = append(turns, Turn{Kind: TurnText, Role: msg.Role, Text: strContent})
			}
			continue
		}

		// Try array of content blocks
		var blocks []json.RawMessage
		if err := json.Unmarshal(msg.Content, &blocks); err != nil {
			continue
		}

		var pendingText string
		for _, rawBlock := range blocks {
			var blockType struct {
				Type string `json:"type"`
			}
			if err := json.Unmarshal(rawBlock, &blockType); err != nil {
				continue
			}

			switch blockType.Type {
			case "text":
				var b struct {
					Text string `json:"text"`
				}
				if json.Unmarshal(rawBlock, &b) == nil && strings.TrimSpace(b.Text) != "" {
					if pendingText != "" {
						pendingText += "\n"
					}
					pendingText += b.Text
				}

			case "tool_use":
				if pendingText != "" {
					turns = append(turns, Turn{Kind: TurnText, Role: msg.Role, Text: pendingText})
					pendingText = ""
				}
				var b struct {
					ID    string          `json:"id"`
					Name  string          `json:"name"`
					Input json.RawMessage `json:"input"`
				}
				if json.Unmarshal(rawBlock, &b) == nil {
					toolNames[b.ID] = b.Name
					inputStr, _ := json.MarshalIndent(mustUnmarshalAny(b.Input), "", "  ")
					turns = append(turns, Turn{
						Kind:     TurnToolCall,
						ID:       b.ID,
						ToolName: b.Name,
						Input:    string(inputStr),
					})
				}

			case "tool_result":
				if pendingText != "" {
					turns = append(turns, Turn{Kind: TurnText, Role: msg.Role, Text: pendingText})
					pendingText = ""
				}
				var b struct {
					ToolUseID string          `json:"tool_use_id"`
					Content   json.RawMessage `json:"content"`
				}
				if json.Unmarshal(rawBlock, &b) == nil {
					content := extractResultContent(b.Content)
					turns = append(turns, Turn{
						Kind:      TurnToolResult,
						ToolUseID: b.ToolUseID,
						ToolName:  toolNames[b.ToolUseID],
						Content:   content,
					})
				}
			}
		}
		if pendingText != "" {
			turns = append(turns, Turn{Kind: TurnText, Role: msg.Role, Text: pendingText})
		}
	}
	return turns
}

func mustUnmarshalAny(raw json.RawMessage) any {
	var v any
	json.Unmarshal(raw, &v)
	return v
}

func extractResultContent(raw json.RawMessage) string {
	// Try string
	var s string
	if json.Unmarshal(raw, &s) == nil {
		return s
	}
	// Try array of text blocks
	var blocks []struct {
		Type string `json:"type"`
		Text string `json:"text"`
	}
	if json.Unmarshal(raw, &blocks) == nil {
		var parts []string
		for _, b := range blocks {
			if b.Type == "text" {
				parts = append(parts, b.Text)
			}
		}
		return strings.Join(parts, "\n")
	}
	return string(raw)
}
